### PR TITLE
Feature/grouping sources

### PR DESCRIPTION
**Pull request by francoisph**

### CODE REVIEW

**Comment by francoisph** [comment]
Functionalities for the source groups :
- Add and remove groups
- Move group
- Move quiddities in another group
- Rename group

File: client/src/view/scenic/pages/matrix/common/ShmdataView.js [hunk 1]
```diff
@@ -55,27 +55,30 @@ const ShmdataView = Marionette.LayoutView.extend( {
...
-            collection: this.options.table.getContactDestinationCollection()
-        } ) );
-
-        this.showChildView( 'rtp', new RTPConnectionsView( {
-            scenic:     this.options.scenic,
-            table:      this.options.table,
-            source:     this.model,
-            collection: this.options.table.getRTPDestinationCollection()
-        } ) );
+        if ( !this.options.table.getSinkGroups().isEmptyGroups() ) {
```
Comment (ubald) [inline client/src/view/scenic/pages/matrix/common/ShmdataView.js]: collectionview is already "hide when empty", you just need the right css to hide it

File: client/src/view/scenic/pages/matrix/common/ShmdataView.js [hunk 2]
```diff
@@ -55,27 +55,30 @@ const ShmdataView = Marionette.LayoutView.extend( {
...
+            this.showChildView( 'sink', new SinkConnectionsView( {
+                scenic:     this.options.scenic,
+                table:      this.options.table,
+                source:     this.model,
+                collection: this.options.table.getSinkGroups()
+            } ) );
+        }
+        if ( !this.options.table.getContactGroups().isEmptyGroups() ) {
```
Comment (ubald) [inline client/src/view/scenic/pages/matrix/common/ShmdataView.js]: collectionview is already "hide when empty", you just need the right css to hide it

File: client/src/view/scenic/pages/matrix/common/ShmdataView.js [hunk 3]
```diff
@@ -55,27 +55,30 @@ const ShmdataView = Marionette.LayoutView.extend( {
...
+            this.showChildView( 'contacts', new ContactConnectionsView( {
+                scenic:     this.options.scenic,
+                table:      this.options.table,
+                source:     this.model,
+                collection: this.options.table.getContactGroups()
+            } ) );
+        }
+        if ( !this.options.table.getRTPGroups().isEmptyGroups() ) {
```
Comment (ubald) [inline client/src/view/scenic/pages/matrix/common/ShmdataView.js]: collectionview is already "hide when empty", you just need the right css to hide it

File: client/src/view/scenic/pages/matrix/contacts/ContactDestinationsView.js [hunk 4]
```diff
@@ -1,32 +1,51 @@
...
-        return this.options.table.contactDestinationComparator( destination );
+    onBeforeShow: function () {
+        this.showChildView( 'groupslist', new DestinationGroupsView( {  scenic:     this.scenic,
+                                                                        table:      this.options.table,
+                                                                        collection: this.collection,
+                                                                        onUpdate:   this.options.onUpdate } ) );
+    },
+ 
+    showGroups: function () {
+        if( this.isEmpty() ) {
```
Comment (ubald) [inline client/src/view/scenic/pages/matrix/contacts/ContactDestinationsView.js]: you can probably remove a lot of handling of empty groups by using the "hideWhenEmpty" behavior
```javascript
    behaviors: {
        HideWhenEmpty: { behaviorClass: HideWhenEmpty }
    }
```
and implementing the right CSS rules

File: client/src/view/scenic/pages/matrix/rtp/RTPDestinationsView.js [hunk 5]
```diff
@@ -1,24 +1,50 @@
...
+    },
+
+    onBeforeShow: function () {
+        this.showChildView( 'groupslist', new DestinationGroupsView( {  scenic:     this.scenic,
+                                                                        table:      this.options.table,
+                                                                        collection: this.collection,
+                                                                        onUpdate:   this.options.onUpdate  } ) );
+    },
+
+    showGroups: function () {
```
Comment (ubald) [inline client/src/view/scenic/pages/matrix/rtp/RTPDestinationsView.js]: idem

File: client/src/view/scenic/pages/matrix/sink/SinkDestinationsView.js [hunk 6]
```diff
@@ -11,19 +11,41 @@ import SinkDestinationView from './SinkDestinationView';
...
+                                                            collection: this.collection,
+                                                            onUpdate:   this.options.onUpdate  } );
+        this.showChildView( 'groupslist', this.destinationGroup  );
+    },
+    
+    addGroup: function ( ) {
+        this.collection.addGroup( "sink" );
+    },
+
+    showGroups: function () {
```
Comment (ubald) [inline client/src/view/scenic/pages/matrix/sink/SinkDestinationsView.js]: IDEM

File: switcher/src/switcher-controller.cpp [hunk 7]
```diff
@@ -144,6 +145,8 @@ void SwitcherController::Init(Local<Object> exports) {
   NODE_SET_PROTOTYPE_METHOD(tpl, "get_info", GetInfo);
   NODE_SET_PROTOTYPE_METHOD(tpl, "get_user_data", GetUserData);
   NODE_SET_PROTOTYPE_METHOD(tpl, "set_user_data", SetUserData);
+  NODE_SET_PROTOTYPE_METHOD(tpl, "set_user_data_branch", SetUserDataBranch);
```
Comment (ubald) [inline switcher/src/switcher-controller.cpp]: update the count in the comment (for us humans) and in ```  tpl->InstanceTemplate()->SetInternalFieldCount(36);
``` (for the machine)

**Comment by francoisph** [comment]
@ubald code review update

File: client/src/model/Groups.js [hunk 8]
```diff
@@ -2,366 +2,169 @@
...
-
-        return result;
-    }, 
-
-    isItemInGroup: function ( groupName, itemId ) {
-        var result = false;
-        var groups = this.getGroups();
-        if ( groups[groupName] && groups[groupName].items ) {
-            result = this._isItemInGroup(groups[groupName].items, itemId);  
+    setItemsDefaultGroup: function() {
```
Comment (ubald) [inline client/src/model/Groups.js]: Document what this does the name is a bit strange to me

File: client/src/model/Groups.js [hunk 9]
```diff
@@ -2,366 +2,169 @@
...
-            }
+    userDataChanged: function( groups ) {
+        if ( groups ) {
+            _.each( groups, group => {
+                var groupAdded = new Group ( null, {id :  group.id,
+                                                    order:  group.order,
+                                                    groupType:   this.groupType,
+                                                    scenic: this.scenic,
+                                                    items: group.items });
+                Backbone.Collection.prototype.add.call( this, groupAdded );
```
Comment (ubald) [inline client/src/model/Groups.js]: I feel strange when I see this everywhere, maybe wrap in an `_add()` method and use this once.

File: client/src/view/scenic/pages/matrix/common/SourceGroupView.js [hunk 10]
```diff
@@ -196,25 +200,21 @@ const SourceGroupView = Marionette.CompositeView.extend( {
...
-        } );
+        this.groups.renameGroup( this.model.id, this.ui.inputName.val());
         this.model.set('editName', false );     
     },
 
     displayInput: function( event ) {
         if ( this.model.id != 'defaultGroup' ) {
             this.model.set('editName', true );
-            this.ui.inputName.focus(); 
+            // this.ui.inputName.focus();
```
Comment (ubald) [inline client/src/view/scenic/pages/matrix/common/SourceGroupView.js]: yes focus and if possible selectAll

File: client/src/model/Groups.js [hunk 11]
```diff
@@ -2,366 +2,169 @@
...
-        var groups = result;
-
-        _.each( groups,  ( data, groupname ) => {
+    reorder: function ( orderGroups ) {
+        this.each( ( model ) => {
             _.each( orderGroups,  ( orderGroup ) => {
-                if ( groupname == orderGroup.group ) {
-                    result[groupname]['order'] = orderGroup.order;
+                if ( model.id == orderGroup.group ) {
+                    model.setOrder( orderGroup.order );
```
Comment (ubald) [inline client/src/model/Groups.js]: you could save here

File: client/src/model/Groups.js [hunk 12]
```diff
@@ -2,366 +2,169 @@
...
-        }        
-
-        // get the value for the oldName and check if the newname is not already used
-       _.each( groups, (data , groupname) => {
-            if (groupname === oldName) {
-                order = data.order;
-            } else if (groupname === newName) {
-                alreadyUsed = true;
-            }
+        this.sortBy(( group ) => {
```
Comment (ubald) [inline client/src/model/Groups.js]: You already have a comparator, this should not be needed

File: client/src/view/scenic/pages/matrix/sink/SinkConnectionsView.js [hunk 13]
```diff
@@ -19,7 +19,16 @@ const SinkConnectionsView = ConnectionsGroupView.extend({
 
     viewComparator: function ( group ) {
         return group.order;
-    }
+    },
+
+    // onRemder: function() {
```
Comment (ubald) [inline client/src/view/scenic/pages/matrix/sink/SinkConnectionsView.js]: onReindeer?

File: server/src/settings/config.js [hunk 14]
```diff
@@ -52,6 +52,30 @@ const config = {
         quiddName: 'defaultRtp'
     },
 
+/*    userTreeQuiddity: {
```
Comment (ubald) [inline server/src/settings/config.js]: remove comment

File: client/src/model/Group.js [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+                result = true;
+            }
+        });
+
+        return result;
+    },
+
+    userDataChanged: function( group ) {
+        if ( group ) {
+            if( this.get('items') !== group.items ) {
```
Comment (francoisph) [inline client/src/model/Group.js]: I want to keep the set items silent, I have some useless refresh on group when I add or remove items

**Comment by francoisph** [comment]
@ubald code review update

**Pull request merged**